### ISSUE TITLE
Segfault with wasm-opt -Oz (and s, 3, 2, 1) - EMSDK 3.1.9

### ISSUE THREAD

**Issue by haberbyte** [WebAssembly/binaryen#4608]
I noticed that when upgrading to emscripten version 3.1.9, the wasm-opt step is failing with a segfault.

It is **failing** with this version
    $ ./vendor/emsdk/upstream/bin/wasm-opt --version
    wasm-opt version 106 (version_101-567-g68807a4f8)

However, with emscripten 3.1.8 it seems to bundle a slightly older version that works.

It is working with this version
    $ ./vendor/emsdk/upstream/bin/wasm-opt --version
    wasm-opt version 106 (version_101-537-g22d24fda9)

I am running this on an M1 Mac, have not tested on others yet.
The error happens with all -O flags (except -O0 which basically does nothing).

**Comment by haberbyte** [comment]
I compiled the recent main branch myself and the issue does not happen

**Comment by kripken** [comment]
Perhaps a stack trace of the crash could help diagnose it?

Also if you can provide the crashing file that would be good. (Can get intermediate emcc files with `EMCC_DEBUG=1` in the env.)

**Comment by DoDoENT** [comment]
> Perhaps a stack trace of the crash could help diagnose it?

@kripken , maybe [this is the stack trace you are looking for](https://github.com/emscripten-core/emscripten/issues/16826#issue-1217522319)?

**Comment by kripken** [comment]
Interesting, if this is the same issue as #16826 then it suggests something is wrong with the emsdk M1 builds, but local M1 builds work - so something is wrong with the emsdk builds. cc @sbc100

**Comment by sbc100** [comment]
I guess its some kind of issue with the version of XCode or the compiler we use to compile the M1 emsdk.  Did something change in binaryen recently that means we use exceptions where we previously didn't.

Sadly, without any actual M1 hardware its hard for @dschuff or @kripken or myself to investigate this kind of thing.  @dschuff any ideas on what we could try changing on our build machine?

**Comment by kripken** [comment]
I don't think we changed anything regarding exceptions for quite some time. I'd guess this might be a new issue only because the M1 builds are new? @haberbyte @DoDoENT did the M1 builds ever work for you before?

**Comment by DoDoENT** [comment]
>did the M1 builds ever work for you before?

Yes, I'm happily using emscripten on M1 since emscripten v3.1.1 (earlier versions required rosetta, afaik). The problems started with 3.1.9, I can normally use 3.1.8 on M1 mac.

**Comment by DoDoENT** [comment]
> it suggests something is wrong with the emsdk M1 builds, but local M1 builds work 

Please see [this comment](https://github.com/emscripten-core/emscripten/issues/16826#issuecomment-1111902834) regarding the build. Even if I make a local build that works and then move the binary to the emscripten SDK, it stops working, but with different error - I don't get a segfault or crash - it just prints that it cannot launch the binary without more context why.

**Comment by kripken** [comment]
Interesting. Ok, as it's a regression then please bisect on the `3.1.8`..`3.1.9` range. Hopefully that will find something useful.

https://emscripten.org/docs/contributing/developers_guide.html#bisecting

However, I don't understand how moving a working build into the emsdk makes it stop working, that's very odd... we just invoke them using python. But perhaps bisection will find that we changed something in how we do that?

**Comment by haberbyte** [comment]
The issue does not happen anymore with 3.1.10.
I think this can be closed?

**Comment by dschuff** [comment]
Just for the record, I mentioned in https://github.com/emscripten-core/emscripten/issues/16826#issuecomment-1112658660 that the underlying cause isn't really fixed, so I'm keeping the emscripten-side bug open. But there's no need to also have this one so we can leave it closed. Thanks for reporting!
